### PR TITLE
Updated the TestStand sequence of NI DCPower measurement to make it simpler

### PR DESCRIPTION
**Pull request by Tharun-Sundar**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
- Updates the TestStand sequence in the DC-Power example measurement to make it simpler by moving the process setup and process cleanup steps to the main sequence.
- Combines the logic for registering pin map into a single step in the Setup section. Previously this was handled in three steps.

### Why should this Pull Request be merged?
Based on the feedback discussed in this [Teams post](https://teams.microsoft.com/l/message/19:36f0e963feb84c0c8122e6ddfbe7df7b@thread.tacv2/1676494204834?tenantId=87ba1f9a-44cd-43a6-b008-6fdb45a5204e&groupId=a092c504-be1c-4d21-9ef4-068a40ca6974&parentMessageId=1676494204834&teamName=MeasurementLink%20%26%20ETW%20Global%20Collaboration&channelName=Roadmap%20Input%20and%20Feature%20Ideas&createdTime=1676494204834&allowXTenantAccess=false), moving the process setup and process cleanup steps to the main sequence will simplify the example sequence as the user will only need to run the main sequence to get the expected result. Though this sequence is not efficient for multi-site use case, this change is encouraged as the example sequence is expected to be tailored to users who are beginners in TestStand.

### What testing has been done?
Manually executed the TestStand example sequence with a simulated DC-Power instrument and verified the result.

### CODE REVIEW

**Review by dixonjoel — APPROVED** [review]

**Comment by Tharun-Sundar** [comment]
@bkeryan When executing the TestStand sequence, the pin map file is not found even if it is available in the sequence directory. I updated the code as shown below to make sure that the logic is similar to the LabVIEW example sequence.
```
sequence_file = self._sequence_context.SequenceFile
(_, file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(file_name, searchContext=sequence_file)
return file_path
```
But it is not having any impact in the search. Please share your thoughts on this.

File: examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq:1]: Please configure the Python adapter consistently for the setup/cleanup steps. The other sequences uncheck **Use Adapter Settings for Python Interpreter** in order to use the example program's virtualenv. They also explicitly specify Python 3.9.

image

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 1]
```diff
@@ -108,3 +108,40 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+
+        """
+        self._sequence_context.Engine.TemporaryGlobals.SetValString(
+            "NI.MeasurementLink.PinMapId", 0x1, pin_map_id
+        )
+
+    def get_file_path(self, file_name: str) -> str:
+        """Return the full path of the input file if it is found in TestStand search directories.
+
+        Args:
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/_helpers.py:144]: Added Returns block

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 2]
```diff
@@ -108,3 +108,40 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+            file_name (str):
+                Name of the file to be found from the TestStand search diectories.
+
+        """
+        (_, file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(file_name)
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Added `searchContext` to specify sequence file path.

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 3]
```diff
@@ -108,3 +108,40 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
                 service_class="ni.measurementlink.v1.grpcdeviceserver",
             ).insecure_address
         )
+
+
+class TestStandSupport(object):
+    """Class that communicates with TestStand."""
+
+    def __init__(self, sequence_context) -> None:
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Annotated the type as `Any` for `sequence_context` COM Object

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 4]
```diff
@@ -48,14 +48,14 @@ def __init__(self, *, grpc_channel: grpc.Channel):
             pin_map_service_pb2_grpc.PinMapServiceStub(grpc_channel)
         )
 
-    def update_pin_map(self, pin_map_id: str) -> None:
+    def update_pin_map(self, pin_map_id: str) -> str:
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Updated

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 5]
```diff
@@ -48,14 +48,14 @@ def __init__(self, *, grpc_channel: grpc.Channel):
...
         """Update registered pin map contents.
 
         Create and register a pin map if a pin map resource for the specified pin map id is not
         found.
 
         Args:
-            pin_map_id (str): The resource id of the pin map to register as a pin map resource. By
+            pin_map_id (str): The file path of the pin map to register as a pin map resource. By
                 convention, the pin map id is the .pinmap file path.
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Added Returns block

File: examples/nidcpower_source_dc_voltage/teststand_fixture.py [hunk 6]
```diff
@@ -1,25 +1,32 @@
...
-            The resource id of the pin map to register as a pin map resource. By
-            convention, the pin_map_id is the .pinmap file path.
+        pin_map_path (str):
+            The path of the pin map to register as a pin map resource.
+        sequence_context:
+            The sequence context object from the TestStand sequence execution.
 
     """
+    teststand_support = TestStandSupport(sequence_context)
+    pin_map_full_path = teststand_support.get_file_path(pin_map_path)
```
[the remark this answers is not in the document]
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/teststand_fixture.py]: Renamed to `pin_map_abs_path`

File: examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq
Comment (vigkre) [reply examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq:1]: Configured Python adapter settings for setup/cleanup steps. Updated Python version to 3.9

**Comment by bkeryan** [comment]
> @bkeryan When executing the TestStand sequence, the pin map file is not found even if it is available in the sequence directory. I updated the code as shown below to make sure that the logic is similar to the LabVIEW example sequence.
> 
> ```
> sequence_file = self._sequence_context.SequenceFile
> (_, file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(file_name, searchContext=sequence_file)
> return file_path
> ```
> 
> But it is not having any impact in the search. Please share your thoughts on this.

When you call `FindFileEx`, the `win32com` module dynamically generates this code:
> `"def FindFileEx(self, fileToFind=pythoncom.Missing, absolutePath=pythoncom.Missing, srchDirType=pythoncom.Missing, searchDirectoryIndex=pythoncom.Missing\n\t\t\t, userCancelled=pythoncom.Missing, promptOption=1, srchListOption=1, isCommand=False, searchContext=pythoncom.Missing\n\t\t\t, reserved=pythoncom.Missing):\n\treturn self._ApplyTypes_(1050, 1, (11, 0), ((8, 1), (16392, 2), (16387, 2), (16387, 2), (16395, 2), (3, 49), (3, 49), (11, 49), (12, 17), (12, 17)), 'FindFileEx', None,fileToFind\n\t\t\t, absolutePath, srchDirType, searchDirectoryIndex, userCancelled, promptOption\n\t\t\t, srchListOption, isCommand, searchContext, reserved)\n"`

When you call `FindFileEx(file_name, searchContext=sequence_file)`, this leaves `absolutePath`, `srchDirType`, etc. unspecified, so they default to `pythoncom.Missing`.

The problem is that `PyIDispatch`  stops passing parameters when it encounters `pythoncom.Missing`: https://github.com/mhammond/pywin32/blob/main/com/win32com/src/PyIDispatch.cpp#L369

Passing `None` for the output parameters seems to work. I initially tried it with positional arguments:
```
        (_, file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(
            file_name, # fileToFind
            None, # absolutePath
            None, # srchDirType
            None, # searchDirectoryIndex
            None, # userCancelled
            1, # promptOption=FindFile_PromptHonorUserPreference
            1, # srchListOption=FindFile_SrchAskUser
            False, # isCommand
            self._sequence_context.SequenceFile, # searchContext
        )
        return file_path
```

Using keyword arguments allows us to leave `promptOption`, `srchListOption`, and `isCommand` unspecified. Their default values are defined in TestStand's IDL, so leaving them unspecified does the right thing.

Let's do this:
```
        (_, file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(
            fileToFind=file_name,
            absolutePath=None,
            srchDirType=None,
            searchDirectoryIndex=None,
            userCancelled=None,
            searchContext=self._sequence_context.SequenceFile,
        )
        return file_path
```

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py:119]: I was referring to the COM object parameter, not the return value.

I will push a couple of commit to update the type annotations and docstrings.

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 7]
```diff
@@ -108,3 +111,57 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+            str: Specifies Absolute Path of the file.
+
+        """
+        sequence_file = self._sequence_context.SequenceFile
+        (_, abs_file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(
+            file_path, searchContext=sequence_file
+        )
+        return abs_file_path
+
+    def get_pin_map_id_temporary_variable(self) -> str:
```
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/_helpers.py:158]: Get and set should be symmetric.

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 8]
```diff
@@ -108,3 +111,57 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+        Args:
+            pin_map_id (str):
+                The resource ID of the pin map that is registered to the pin map service.
+
+        """
+        self._sequence_context.Engine.TemporaryGlobals.SetValString(
+            "NI.MeasurementLink.PinMapId", 0x1, pin_map_id
+        )
+
+    def get_file_path(self, file_path: str) -> str:
```
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/_helpers.py:141]: This should be named resolve like in LV.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 9]
```diff
@@ -108,3 +111,57 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+            file_path (str):
+                Name of the file to be found from the TestStand search directories.
+
+        Returns:
+            str: Specifies Absolute Path of the file.
+
+        """
+        sequence_file = self._sequence_context.SequenceFile
+        (_, abs_file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(
+            file_path, searchContext=sequence_file
```
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/_helpers.py:153]: This seems to return "." when you pass in an absolute file path or you click cancel on the file dialog, which is not what we want.

**Review by bkeryan — CHANGES_REQUESTED** [review]

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 10]
```diff
@@ -108,3 +111,55 @@ def get_grpc_device_channel(self, provided_interface: str) -> grpc.Channel:
...
+        Args:
+            file_path:
+                An absolute or relative path to the file. If this is a relative path, this function
+                searches the TestStand search directories for it.
+
+        Returns:
+            An absolute path to the file.
+        """
+        sequence_file = self._sequence_context.SequenceFile
+        (_, abs_file_path, _, _, _) = self._sequence_context.Engine.FindFileEx(
```
[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py:152]: I suppose this isn't a new issue, though. And I don't see any way to make FindFileEx behave differently. Marking resolved.

File: examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq:1]: Update Pin Map still has "Use Adapter Settings for Python Interpreter" checked.

image

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 7 above)
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Fixed

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 9 above)
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Fixed:
- Check for absolute paths before calling FindFileEx
- Check userCancelled output

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Fixed

File: examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/NIDCPowerSourceDCVoltage.seq:1]: Fixed

File: examples/nidcpower_source_dc_voltage/_helpers.py [hunk 11]
```diff
@@ -48,23 +48,26 @@ def __init__(self, *, grpc_channel: grpc.Channel):
...
-            pin_map_id (str): The resource id of the pin map to register as a pin map resource. By
+            pin_map_path (str): The file path of the pin map to register as a pin map resource. By
                 convention, the pin map id is the .pinmap file path.
 
+        Returns:
+            str: Specifies the registered pin map id.
+
         """
-        pin_map_path = pathlib.Path(pin_map_id)
+        pin_map_purepath = pathlib.Path(pin_map_path)
```
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/_helpers.py:65]: This is not a pathlib.PurePath. pathlib.PurePath is like pathlib.Path except it doesn't support filesystem access, and this function performs filesystem access.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 11 above)
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Fixed

**Review by bkeryan — APPROVED** [review]

File: examples/nidcpower_source_dc_voltage/_helpers.py (hunk 11 above)
Comment (bkeryan) [reply examples/nidcpower_source_dc_voltage/_helpers.py]: Renamed

**Pull request merged**